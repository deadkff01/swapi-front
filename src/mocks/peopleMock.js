export const peopleMock = [
  {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male',
    homeworld: 'http://swapi.dev/api/planets/1/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/6/'
    ],
    species: [],
    vehicles: [
      'http://swapi.dev/api/vehicles/14/',
      'http://swapi.dev/api/vehicles/30/'
    ],
    starships: [
      'http://swapi.dev/api/starships/12/',
      'http://swapi.dev/api/starships/22/'
    ],
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    url: 'http://swapi.dev/api/people/1/'
  },
  {
    name: 'C-3PO',
    height: '167',
    mass: '75',
    hair_color: 'n/a',
    skin_color: 'gold',
    eye_color: 'yellow',
    birth_year: '112BBY',
    gender: 'n/a',
    homeworld: 'http://swapi.dev/api/planets/1/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/4/',
      'http://swapi.dev/api/films/5/',
      'http://swapi.dev/api/films/6/'
    ],
    species: ['http://swapi.dev/api/species/2/'],
    vehicles: [],
    starships: [],
    created: '2014-12-10T15:10:51.357000Z',
    edited: '2014-12-20T21:17:50.309000Z',
    url: 'http://swapi.dev/api/people/2/'
  },
  {
    name: 'R2-D2',
    height: '96',
    mass: '32',
    hair_color: 'n/a',
    skin_color: 'white, blue',
    eye_color: 'red',
    birth_year: '33BBY',
    gender: 'n/a',
    homeworld: 'http://swapi.dev/api/planets/8/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/4/',
      'http://swapi.dev/api/films/5/',
      'http://swapi.dev/api/films/6/'
    ],
    species: ['http://swapi.dev/api/species/2/'],
    vehicles: [],
    starships: [],
    created: '2014-12-10T15:11:50.376000Z',
    edited: '2014-12-20T21:17:50.311000Z',
    url: 'http://swapi.dev/api/people/3/'
  }
];

export const C3PO = {
  name: 'C-3PO',
  height: '167',
  mass: '75',
  hair_color: 'n/a',
  skin_color: 'gold',
  eye_color: 'yellow',
  birth_year: '112BBY',
  gender: 'n/a',
  homeworld: 'http://swapi.dev/api/planets/1/',
  films: [
    'http://swapi.dev/api/films/1/',
    'http://swapi.dev/api/films/2/',
    'http://swapi.dev/api/films/3/',
    'http://swapi.dev/api/films/4/',
    'http://swapi.dev/api/films/5/',
    'http://swapi.dev/api/films/6/'
  ],
  species: ['http://swapi.dev/api/species/2/'],
  vehicles: [],
  starships: [],
  created: '2014-12-10T15:10:51.357000Z',
  edited: '2014-12-20T21:17:50.309000Z',
  url: 'http://swapi.dev/api/people/2/'
};
