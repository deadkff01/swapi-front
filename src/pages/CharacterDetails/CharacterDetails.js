import { useEffect, useState, useCallback } from 'react';
import { Container } from 'components/Container';
import { Title } from 'components/Title';
import { useHttpInstance } from 'services/baseService';
import { useParams } from '@reach/router';
import { Loader } from 'components/Loader';
import { ErrorWarning } from 'components/ErrorWarning';
import { Fade } from 'utils/Fade';
import { BackButton } from 'components/BackButton';

export default function CharacterDetails() {
  const [character, setCharacter] = useState({});
  const params = useParams();
  const { get, response, error, loading } = useHttpInstance();

  const getCharacter = useCallback(async () => {
    const people = await get(`people/${params.id}`);
    if (response.ok) setCharacter(people);
  }, [get, response, params.id]);

  useEffect(() => {
    getCharacter();
  }, [getCharacter]);

  if (loading) {
    return <Loader />;
  }

  if (error) {
    return (
      <ErrorWarning message="Error, verify your conection or reload the page! may the force be with you!" />
    );
  }

  return (
    <Fade in={true} unmountOnExit appear>
      <Container
        data-testid="container"
        className="flex flex-wrap justify-center "
      >
        <BackButton />
        <div className="w-full text-center px-5 mb-5">
          <Title className="text-4xl" align="center">
            {character.name}
          </Title>

          <div className="flex mb-5 justify-center">
            <div className="text-center text-lg px-5">
              <Title className="text-3xl" align="center">
                Character Details
              </Title>
              <div>Name: {character.name}</div>
              <div>Height: {character.height}</div>
              <div>Mass: {character.mass}</div>
              <div>Gender: {character.gender}</div>
              <div>Skin Color: {character.skin_color}</div>
              <div>Hair Color: {character.hair_color}</div>
            </div>
          </div>
        </div>
      </Container>
    </Fade>
  );
}
