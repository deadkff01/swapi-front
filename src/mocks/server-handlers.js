import { rest } from 'msw';
import { mockMovies, aNewHopeMock } from './mockMovies';
import { peopleMock, C3PO } from './peopleMock';

const handlers = [
  rest.get(`${process.env.REACT_APP_ENDPOINT}/movies`, (_, res, ctx) =>
    res(ctx.json(mockMovies))
  ),
  rest.get(`${process.env.REACT_APP_ENDPOINT}/movies/1`, (_, res, ctx) =>
    res(ctx.json(aNewHopeMock))
  ),
  rest.post(`${process.env.REACT_APP_ENDPOINT}/people`, (_, res, ctx) =>
    res(ctx.json(peopleMock))
  ),
  rest.get(`${process.env.REACT_APP_ENDPOINT}/people/2`, (_, res, ctx) =>
    res(ctx.json(C3PO))
  )
];

export { handlers };
