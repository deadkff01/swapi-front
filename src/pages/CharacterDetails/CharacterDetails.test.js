import { cleanup, screen } from '@testing-library/react';
import '../../setupTests';
import { renderWithRouterParams } from '../../testUtils';
import CharacterDetails from './CharacterDetails';
import '@testing-library/jest-dom/extend-expect';
import { server } from 'mocks/server';

describe('CharacterDetails component', () => {
  beforeAll(() => server.listen());

  afterEach(() => server.resetHandlers());

  afterAll(() => {
    jest.useRealTimers();
    server.close();
    cleanup();
  });

  test('renders the component', async () => {
    renderWithRouterParams(<CharacterDetails path="/character/:id" />, {
      route: '/character/2'
    });
    const container = await screen.findByTestId('container');
    expect(container).toMatchSnapshot();
  });

  test('renders the loader', async () => {
    renderWithRouterParams(<CharacterDetails path="/character/:id" />, {
      route: '/character/2'
    });
    const loader = await screen.findByTestId('loader');
    expect(loader).toBeDefined();
  });

  test('renders with valid values', async () => {
    renderWithRouterParams(<CharacterDetails path="/character/:id" />, {
      route: '/character/2'
    });
    const name = await screen.findByText('C-3PO');
    const height = await screen.findByText('Height: 167');
    const mass = await screen.findByText('Mass: 75');
    const gender = await screen.findByText('Gender: n/a');
    const skinColor = await screen.findByText('Skin Color: gold');
    const hairColor = await screen.findByText('Hair Color: n/a');

    expect(name).toBeDefined();
    expect(height).toBeDefined();
    expect(mass).toBeDefined();
    expect(gender).toBeDefined();
    expect(skinColor).toBeDefined();
    expect(hairColor).toBeDefined();
  });

  // test('renders the error message', async () => {
  //   server.use(
  //     rest.get(`${process.env.REACT_APP_ENDPOINT}/movies`, (_, res, ctx) =>
  //     res(ctx.json('ddd'))
  //   )
  //     // rest.get('http://localhost:8080/movies', async (req, res, ctx) => {
  //     //   return res(
  //     //     ctx.status(500),
  //     //     ctx.json({ message: 'Internal Server Error' })
  //     //   );
  //     // })
  //   );
  //   renderWithRouter(<Movies />);
  //   const error = await screen.findByTestId('error-message');
  //   expect(error).toBeDefined();
  // });
});
