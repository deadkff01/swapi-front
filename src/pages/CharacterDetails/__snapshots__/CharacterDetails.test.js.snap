// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetails component renders the component 1`] = `
<div
  class="sc-bdfBwQ bqIkDZ flex flex-wrap justify-center "
  data-testid="container"
>
  <button
    class="sc-eCssSg cgjOyw mt-5"
    data-testid="movie-card"
  >
    &lt; 
     Back
  </button>
  <div
    class="w-full text-center px-5 mb-5"
  >
    <h1
      class="sc-gsTCUz ipzQyr text-4xl"
    />
    <div
      class="flex mb-5 justify-center"
    >
      <div
        class="text-center text-lg px-5"
      >
        <h1
          class="sc-gsTCUz ipzQyr text-3xl"
        >
          Character Details
        </h1>
        <div>
          Name: 
        </div>
        <div>
          Height: 
        </div>
        <div>
          Mass: 
        </div>
        <div>
          Gender: 
        </div>
        <div>
          Skin Color: 
        </div>
        <div>
          Hair Color: 
        </div>
      </div>
    </div>
  </div>
</div>
`;
